### ISSUE TITLE
Deprecate CONTENTdm .cpd files in favor of generic structure files

### ISSUE THREAD

**Issue by mjordan** [MarcusBarnes/mik#220]
This issue documents a discussion between myself and @MarcusBarnes about the relationship between the forthcoming Islandora Compound Batch module and various MIK tookchains.

We currently have a CONTENTdm Compound Object toolchain, the development of which is documented in issue #111. The Islandora Compound Batch module parses the .cpd files as described in #111 and creates relationships in the resulting Islandora objects that mirror the structure of the source CONTENTdm object.

Now that we want to accommodate compound objects that do not come from CONTENTdm, and that therefore lack a .cpd file, we will be adopting a way of expressing the structure of a compound object using a generic XML file like the following:

```xml
<?xml version="1.0" encoding="utf-8"?>
<islandora_compound_object title="I am a compound object">
  <child content="The first child"/>
  <child content="The second child"/>
</islandora_compound_object>
```

This structure is simple and flexible enough to accommodate hierarchical relationships, but for now we will be focusing on "flat" structures that only have one level of parent-child relationship. Upstream tools, like MIK's CONTENTdm Compound Object toolchain, will need to be able to generate this type of XML file. We have added a helper script to the  Islandora Compound Batch module that will generate a structure file from a directory structure, and plan to modify the MIK CONTENTdm Compound Object toolchain so that it does not write out its .cpd file. Once we do that, the same script that generates the structure file from a directory structure can be used with the CONTENTdm Compound Object toolchain output.

One implication of using a directory tree as the basis for an XML file that expresses an Islandora compound object's structure is that to accommodate multilevel hierarchical structures, the application (or human) preparing the content for ingest into Islandora will need to represent that structure as a set of nested filesystem directories. Once a structure file is generated from that directory tree, it can be manually edited to correct sorting/sequence/order that should be preserved in the parent-child relationships defined by Islandora, given the vagaries of "sorting" files and directories in graphical file managers and command-line tools like `tree`. 

@MarcusBarnes writing this makes me realize that we'll need to account for situations where the filesystem arrangement and the directory tree disagree. We probably should choose the structure file, for the reason I mention in the previous paragraph.

**Comment by mjordan** [comment]
One issue we need to keep in mind if we are going to have MIK represent the structure of a CONTENTdm compound object as a directory tree is that we need to preserve the order of the child objects. Using the .cpd file below as an example, and assuming we would create child object subdirectories using the value of the `<pageptr>` elements, the second child may precede the first, depending on how the script generating the structure.xml file sorts the contents of the directoy:

```xml
<?xml version="1.0" encoding="utf-8"?>
<cpd>
  <type>Document</type>
  <page>
    <pagetitle>MSC130-12166-01</pagetitle>
    <pagefile>101.jp2</pagefile>
    <pageptr>99</pageptr>
  </page>
  <page>
    <pagetitle>MSC130-12166-02</pagetitle>
    <pagefile>102.jp2</pagefile>
    <pageptr>100</pageptr>
  </page>
</cpd>
```

Perhaps the best way of preventing this from happening is to have MIK use XSLT to transform the CONTENTdm .cdp file into the generic structure.xml file, preserving the sequence of child objects.

**Comment by mjordan** [comment]
We can have MIK produce Islandora compound structure files from CONTENTdm .cpd files using the following code:

```php
<?php

// Usage: php cpd_to_islandora_dom.php cpd_to_islandora.xsl cpd.xml

$xml_path = trim($argv[1]);

$input_dom = new DOMDocument;
$input_dom->preserveWhiteSpace = false;
$input_dom->formatOutput = true;
$input_dom->load($xml_path);

$output_dom = new DOMDocument('1.0', 'utf-8');
$output_dom->preserveWhiteSpace = false;
$output_dom->formatOutput = true;
$root = $output_dom->createElement('islandora_compound_object');
$comment = $output_dom->createComment('Islandora compound structure file generated from a CONTENTdm .cpd file.');
$root->appendChild($comment);

$xpath = new DOMXPath($input_dom);
$pagepointers = $xpath->query("//pageptr");
foreach ($pagepointers as $pageptr) {
  $pageptr_value = $pageptr->nodeValue;
  $child = $output_dom->createElement('child');
  $content = $output_dom->createAttribute('content');
  $content->value = $pageptr_value;
  $child->appendChild($content);
  $root->appendChild($child);
}

$output_dom->appendChild($root);

print $output_dom->saveXML();
```

This preserves the order of the elements in the .cpd file, addressing the issue I raised in the previous comment. Applied to the .cpd file in that comment, this script ouputs:

```xml
<?xml version="1.0" encoding="utf-8"?>
<islandora_compound_object>
  <!--Islandora compound structure file generated from a CONTENTdm .cpd file.-->
  <child content="99"/>
  <child content="100"/>
</islandora_compound_object>
```

Perhaps we can make generation of these files the default, but add a setting to the `[WRITER]` section of CONTENTdm toolchains to not produce this file but to write out the structure.cpd file instead. @MarcusBarnes what do you think?

**Comment by MarcusBarnes** [comment]
@mjordan I think that would be a suitable solution to the issue.  Thank you.

**Comment by mjordan** [comment]
OK,  thanks, I'll integrate this solution into the Cdm writers.

**Comment by mjordan** [comment]
@MarcusBarnes can you test? Run the issue-220 branch against the BC Postcards. Your output should contain `structure.xml` files using thw `<islandora_compound_object>` XML. If you want to retain the CPD file in addition to the new `structure.xml` file, add `cpd_filename = foo.xml` to your .ini file's `[WRITER]` section and the CPD file will be written to a file with the specified name.

I'd like to test this and the latest version of the batch loader that uses the `structure.xml` a bit more before merging.

**Comment by MarcusBarnes** [comment]
@mjordan Yes - I'll test.  Thank you!

**Comment by MarcusBarnes** [comment]
@mjordan I've tested BC Postcard output and the MIK created packages as expected.  I'll double check that the batch loader handle the packages MIK created as expected.

I noticed that we'll need to update the validate_mods.php postwrite hook to handle the CdmCompound case as the default case doesn't work with CdmCompound https://github.com/MarcusBarnes/mik/blob/master/extras/scripts/postwritehooks/validate_mods.php#L53.  The validation failures were handled gracefully however and logged to the postwritehook_validate_mods_error.log file.

**Comment by mjordan** [comment]
Thanks. I hadn't noticed that the post-write hook didn't work for compound objects because I've gotten into the habit of running https://github.com/MarcusBarnes/mik/blob/master/extras/scripts/validate_mods_xml.php to validate MODS, which recurses down a directory tree.

**Comment by MarcusBarnes** [comment]
@mjordan I've tested batch loading the packages I created earlier and everything appears to work as expected.  Please feel free to merge if you feel comfortable with the state of things.

**Comment by mjordan** [comment]
Will do!

**Comment by mjordan** [comment]
Merged into master with 32e0697dbb5a9e4171991a49829ba08c3bdcf204.

Note that there are some ingest packages queued up for import that use .cpd files, not structure.xml files, so the Islandora Compound Batch module should not be updated until these are done.
